### DISCUSSION TITLE
Create users with API

### DISCUSSION THREAD

**Discussion by AntonyKing7D** [drakkan/sftpgo#615] [Q&A] [answered]
Hi, thanks for the excellent product, it's working really well for us.

One thing I would like to be able to do is create (and remove) users via the API. I'm trying to send this to it but it's responding with error 400 - {"error":"EOF","message":""} . Here;s what I'm trying to send:

```
{
  "status": 1,
  "username": "testuser",
  "expiration_date": 0,
  "public_keys": [
    "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEUWwDwEWhTbF0MqAsp/oXK1HR2cElhM8oo1uVmL3ZeDKDiTm4ljMr92wfTgIGDqIoxmVqgYIkAOAhuykAVWBzc= user@host"
  ],
  "filesystem": {
    "provider": 2,
    "gcsconfig": {
      "bucket": "mybucketname",
      "automatic_credentials": 1,
      "storage_class": "default",
      "key_prefix": "testuser/"
    }
  },
  "permissions": {
    "/": [
      "*"
    ]
  },
  "additional_info": "Test account"
}
```
I'm using the automatic credentials to access the bucket, which works when setting up users via the gui. What have I missed ?

Many thanks,

Antony King
7digital.

**Comment by drakkan** [comment, marked as answer]
Hi,

first of all thank you for supporting SFTPGo, much appreciated.

Your request looks fine. I just tested it

```
curl -X 'POST' \
  'http://localhost:8080/api/v2/users' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiQVBJIl0sImV4cCI6MTYzNzU5OTgwNCwianRpIjoiYzZkc2IzNjEwZDFtNzNrdmJxY2ciLCJuYmYiOjE2Mzc1OTg1NzQsInBlcm1pc3Npb25zIjpbIioiXSwic3ViIjoiTzN3ZzFDd0V6RU1uRGRvUFBOQW5hU3JBVjF5c1AyUjRKeDVsV1VBNEpmMD0iLCJ1c2VybmFtZSI6ImFkbWluIn0.TJgx4IihwFKxBuiKxmRdx8ZBxymmoz8DYGrsy0QplLI' \
  -H 'Content-Type: application/json' \
  -d '{
  "status": 1,
  "username": "testuser",
  "expiration_date": 0,
  "public_keys": [
    "ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEUWwDwEWhTbF0MqAsp/oXK1HR2cElhM8oo1uVmL3ZeDKDiTm4ljMr92wfTgIGDqIoxmVqgYIkAOAhuykAVWBzc= user@host"
  ],
  "filesystem": {
    "provider": 2,
    "gcsconfig": {
      "bucket": "mybucketname",
      "automatic_credentials": 1,
      "storage_class": "default",
      "key_prefix": "testuser/"
    }
  },
  "permissions": {
    "/": [
      "*"
    ]
  },
  "additional_info": "Test account"
}'
```

Are you sending a POST request? Can you please share the sftpgo logs?

If you update to the latest development version (that will become v2.2.0 the next weekend). You can test the SFTPGo API using swagger ui at the url `http://<sftpgo_id>:8080/openapi/swagger-ui`. If you paste your request in swagger ui you can see that it works fine

**Reply by AntonyKing7D** [reply]
Perfect, thanks for that, particularly for including the curl that you tested it with. I was using stdin, and had missed the `@-` from the -d flag. :facepalm: 

So my working command is:

 `cat newuser.json | curl -d@-  -i --header 'accept: application/json' -X POST --header "Authorization: Bearer $TOKEN"  --header 'Content-Type: application/json'   my.host.name:8080/api/v2/users`

where newuser.json contains the json as above and $TOKEN is the token fetched from the /api/v2/token endpoint.

Thanks for your help,
Antony.

**Reply by drakkan** [reply]
> Perfect, thanks for that, particularly for including the curl that you tested it with. I was using stdin, and had missed the `@-` from the -d flag. facepalm
> 
> So my working command is:
> 
> `cat newuser.json | curl -d@- -i --header 'accept: application/json' -X POST --header "Authorization: Bearer $TOKEN" --header 'Content-Type: application/json' my.host.name:8080/api/v2/users`
> 
> where newuser.json contains the json as above and $TOKEN is the token fetched from the /api/v2/token endpoint.
> 
> Thanks for your help, Antony.

No problem, thanks for supporting the project. If you have any other issues, don't hesitate to ask
